geode-apis-compatible-with-redis/src/main/java/org/apache/geode/redis/internal/collections/Object2ObjectOpenCustomHashMapWithCursor.java

```diff
@@ -0,0 +1,185 @@
+    super(m, strategy);
+  }
+
+  public Object2ObjectOpenCustomHashMapWithCursor(K[] k, V[] v, float f,
+      Strategy<? super K> strategy) {
+    super(k, v, f, strategy);
+  }
+
+  public Object2ObjectOpenCustomHashMapWithCursor(K[] k, V[] v,
+      Strategy<? super K> strategy) {
+    super(k, v, strategy);
+  }
+
+  /**
+   * Scan entries and pass them to the given consumer function, starting at the passed in
+   * cursor. This method will scan until at least count entries are returned, or the entire
+   * map has been scanned. Once the returned cursor is 0, the entire map is scanned.
+   *
+   * This method may emit more than *count* number of elements if there are hash collisions.
+   *
+   * @param cursor The cursor to start from. Should be 0 for the initial scan. Subsequent calls
+   *        should use the cursor returned by the previous scan call.
+   * @param count The number of elements to scan
+   * @param consumer A function to pass the scanned keys and values to
+   * @param privateData Some data to pass to the function, for example a map to collect values in.
+   *        This
+   *        allows the function to be stateless.
+   * @param <D> The type of the data passed to the function/
+   * @return The next cursor to scan from, or 0 if the scan has touched all elements.
+   */
+  public <D> int scan(int cursor, int count, EntryConsumer<K, V, D> consumer, D privateData) {
+    // Implementation notes
+    //
+    // This stateless scan cursor algorithm is based on the dictScan cursor
+    // implementation from dict.c in redis. Please see the comments in that class for the full
+    // details. That iteration algorithm was designed by Pieter Noordhuis.
+    //
+    // There is one wrinkle due to the fact that we are using a different type of hashtable here.
+    // The parent class, Object2ObjectOpenHashMap, uses an open addressing with a linear
+    // probe. What that means is that when there is a hash collision, instead of putting
+    // a linked list of hash entries into a single hash bucket, this implementation simply
+    // moves on to the next element to the right in the array and tries to put the inserted
+    // object there, continuing until it finds a null slot.
+    //
+    // So in order to use the redis cursor algorithm, our scan needs to probe ahead to
+    // subsequent positions to find any hash entries that match the position we are scanning.
+    // This is logically equivalent to iterating over the linked list in a hashtable bucket
+    // for a redis style closed addressing hashtable.
+    //
+
+    do {
+      // Emit all of the entries at the cursor. This means looking forward in the hash
+      // table for any non-null entries that might hash to the current cursor and emitting
+      // those as well. This may even wrap around to the front of the hashtable.
+      int position = cursor;
+      while (key[position & mask] != null) {
+        if (elementAtHashesTo(position, cursor & mask)) {
```

Comment:
this method name reads kinda weird. maybe `elementAtCursorHashesTo`?

Reply:
I'm going to merge this and create a new PR with a JMH test for this. I'll rename this method in the next PR. Thanks!